### PR TITLE
api: create and delete blueprint endpoint

### PR DESCRIPTION
**Pull request by lzap**
This is my proposal for the create blueprint endpoint. It is 1:1 copy of the payload from the create compose request. Let’s discuss the API first, then I will start working on the code and DDL.

### CODE REVIEW

**Comment by lzap** [comment]
Ping @croissanne any comments before I start?

**Comment by croissanne** [comment]
> Ping @croissanne any comments before I start?

Go for it :)

**Comment by lzap** [comment]
Rebased the code logic for inserting the record into the database. As part of the PR, I am introducing new type named `BlueprintV1` and a version flag that is stored in a separate column. This allows taking advantage of types in Go as I expect us to modify the schema a bit and unstructured data in SQL database is always a problem: existing records on stage will carry old versions which can be easily "upgraded".

Other than that, it is just a copy-paste.

File: internal/db/migrations-tern/007_blueprint_composes.sql
Comment (croissanne) [inline internal/db/migrations-tern/007_blueprint_composes.sql:4]: Does it make sense to enforce the compose_id to be unique in this table?

**Review by croissanne — COMMENTED** [review]
Overall this lgtm

File: internal/db/migrations-tern/007_blueprint_composes.sql
Comment (lzap) [reply internal/db/migrations-tern/007_blueprint_composes.sql:4]: Good point, it does. Now that I am thinking about this table, I am a bit ahead as creating composes through the new experimental API is not part of this PR, but if you don’t mind I will go ahead, add unique constraint and leave it here.

**Comment by lzap** [comment]
Ok rebased and added UNIQUE constraint.

File: internal/v1/api.yaml
Comment (lzap) [inline internal/v1/api.yaml:404]: Ha I should be returning blueprint id instead. Do not merge.

**Comment by lzap** [comment]
So I have also created new endpoint `/blueprints/UUID/compose` which forwards the creation to the existing code. I had to slightly refactor one handler, I introduced `composeImage` which is called from `ComposeImage` as well as `ComposeBlueprint` except the latter uses input data from the database.

File: cmd/image-builder-db-test/main_test.go
Comment (lzap) [inline cmd/image-builder-db-test/main_test.go:399]: Renamed because `TestMain` is reserved for test suites, it does work but Goland is marking this as an error and it is pesky. Also refactored it to function so `defer` can be used to ensure all data is deleted after even an error.

Comment (lzap) [inline cmd/image-builder-db-test/main_test.go:399]: Renamed because TestMain is reserved for test suites, it does work but Goland is marking this as an error and it is pesky. Also refactored it to function so defer can be used to ensure all data is deleted after even an error.

File: internal/tutils/psql.go
Comment (lzap) [inline internal/tutils/psql.go:53]: This significantly speeds up on one of my systems (old Xeon with HDD). From 336s to just 36s.

File: internal/tutils/psql.go
Comment (lzap) [inline internal/tutils/psql.go:94]: The PR contains couple of error printing, during initial setup I had problems (missing `podman` or misconfiguration there) and it was hard to troubleshoot without these.

File: internal/tutils/psql.go
Comment (lzap) [inline internal/tutils/psql.go:132]: This was the culprint - `Output()` is missing standard error completely, so the error was actually hidden.

**Comment by lzap** [comment]
This is now ready for re-review, I removed "compose blueprint" endpoint, this is only create operation (with get/delete DB operations). The rest will be followed in separate PRs. I am leaving few small changes in tests and db tests because I stuggled hard to set it up and with performance. @croissanne

**Comment by lzap** [comment]
Resolved conflicts.

File: internal/v1/api.yaml
Comment (lucasgarfield) [inline internal/v1/api.yaml:643]: This is probably leftover from a copy/paste... shouldn't there be a request per target?

File: internal/v1/api.yaml
Comment (lucasgarfield) [inline internal/v1/api.yaml:638]: We need to support all target types, and that number will probably increase... maybe no max needed?

File: internal/db/migrations-tern/006_create_table_blueprints.sql
Comment (lzap) [reply internal/db/migrations-tern/006_create_table_blueprints.sql:9]: This is not blueprint version from the users perspective, this is body (payload) version which will be increased as soon as incompatible JSON change will be done. This must be stored outside of the JSON itself because only then you can effectively pick the proper type, `BlueprintV1` in this case, and unmarshal data into specific type.

Blueprint versioning will be done in a separate PR and that field will be likely called simply `version`, or we might end up only having `parent uuid` column. We talked about this with @avitova today on techdiss, it needs more research on what will be more appropriate for this case.

File: internal/v1/api.yaml
Comment (lzap) [reply internal/v1/api.yaml:643]: This is copy-pasted from the "create compose" endpoint without changes, the way I understand this is that this only accepts array with exactly one element and this is for the future (e.g. when multiple requests can be send). @croissanne ?

File: internal/v1/api.yaml
Comment (lzap) [reply internal/v1/api.yaml:638]: Ditto.

File: internal/db/migrations-tern/006_create_table_blueprints.sql
Comment (ezr-ondrej) [reply internal/db/migrations-tern/006_create_table_blueprints.sql:9]: Ah! Gotcha, than ignore my comment, this makes sense!

File: internal/v1/api.yaml
Comment (croissanne) [reply internal/v1/api.yaml:643]: So we'd generate multiple compose requests from the blueprint, so this will have to change in future for sure (but we can change stuff under /experimental so).

And then we also might arrive at a point where a single compose requests can have multiple upload targets, which will be the case for edge (s3 + pulp).

Comment (lzap) [reply internal/v1/api.yaml:643]: So if you ask me @lucasgarfield I'd say out of scope for this PR :-D

File: internal/v1/api.yaml
Comment (lzap) [reply internal/v1/api.yaml:638]: I suggest to do this separately after this is merged in both components.

Comment (ezr-ondrej) [reply internal/v1/api.yaml:638]: This should be different in Blueprint and Compose, here it should be 1:n, but compose stay at 1:1, we'd just make N composes when making composes out of blueprint (rebuilding blueprint images) or do I misunderstand something?

I'm not opposed to do it in follow up, I just don't get why 🤔 but sure, I can propose it myself 👍 I can have contribution!! 😛

Comment (lzap) [reply internal/v1/api.yaml:638]: I don’t understand, this is a datatype from "create compose" API request without any change. I believe that blueprint is just a "template" of what we have today. Therefore, I designed this endpoint in a way that blueprint can be resubmitted "as is" without any changes.

Maybe let’s have a chat on Slack to speed things up, I might be missing something.

Comment (lzap) [reply internal/v1/api.yaml]: Oh now I understand, multiple image requests can be in a single blueprint. This is different from what is now. Fixed.

**Comment by lzap** [comment]
Rebased, this is now ready for re-review, thanks.

**Comment by ezr-ondrej** [comment]
The tests seem to not agree with the swagger spec, I guess it needs to get updated? :thinking:

**Comment by lzap** [comment]
Oh right, rebased.

**Comment by ezr-ondrej** [comment]
We are :green_apple: :tada:

**Comment by lzap** [comment]
Let's go @croissanne or @lucasgarfield we have more than one PRs to work on. Thanks.

File: cmd/image-builder/main.go
Comment (croissanne) [inline cmd/image-builder/main.go:5]: external imports tend to go in the 2nd paragraph, it's a bit of a mess in this file though.

File: internal/db/migrations-tern/006_create_table_blueprints.sql
Comment (croissanne) [reply internal/db/migrations-tern/006_create_table_blueprints.sql:9]: JSON unmarshaling is very forgiving though, so you could just grab the version and then unmarshal it again into the appropriate struct.

Once we build versioning on top we need a way to group the blueprints indeed.

File: internal/v1/api.yaml
Comment (croissanne) [reply internal/v1/api.yaml:642]: also I think it's fine to keep this as an array, for koji we ended up accepting multiple image requests per compose reqs so who knows where we need to go with this.

File: internal/db/db_blueprints.go
Comment (croissanne) [inline internal/db/db_blueprints.go:7]: nitpick, space between built-in imports and external imports.

File: internal/tutils/psql.go
Comment (croissanne) [reply internal/tutils/psql.go:94]: would be nice so split out test improvements to a separate commit.

**Review by croissanne — COMMENTED** [review]
Thank you! overal lgtm, some nitpicks

File: cmd/image-builder/main.go
Comment (lzap) [reply cmd/image-builder/main.go:5]: I noticed, we use an automated tool to fix them all `goimports` it does this kind of sorting for us. I guess I will propose a new PR soon.

File: internal/db/db_blueprints.go
Comment (lzap) [reply internal/db/db_blueprints.go:7]: Fixing too.

**Comment by lzap** [comment]
Rebased, I will do a followup on those nasty imports scattered around.

The test refactoring is unfortunately buried deep in the patch, I can do hunk by hunk split if you insist.

**Comment by lzap** [comment]
All green.

**Comment by lzap** [comment]
Here is the followup which I can rebase after this is merged: https://github.com/osbuild/image-builder/pull/895

File: cmd/image-builder/main.go
Comment (croissanne) [reply cmd/image-builder/main.go:5]: right, can you move the gommon/log down though?

File: cmd/image-builder/main.go
Comment (croissanne) [inline cmd/image-builder/main.go:141]: Can you please split this into a separate commit?

Comment (lzap) [reply cmd/image-builder/main.go:141]: Sure, done.

File: cmd/image-builder/main.go
Comment (lzap) [reply cmd/image-builder/main.go]: Done.

**Comment by lzap** [comment]
Some unrelated changes I guess?

/retest

**Comment by ezr-ondrej** [comment]
/retest

> Some unrelated changes I guess?

[quai](https://quay.io/) was  down for the whole day ;)

**Comment by lzap** [comment]
ALL TESTS GREEN.

**Comment by lzap** [comment]
All green.

**Review by croissanne — APPROVED** [review]

**Pull request merged**
